system/jlib/jsocket.cpp

```diff
@@ -4780,14 +4780,44 @@ class CSocketEpollThread: public CSocketBaseThread
     void updateItems()
     {
         // must be in CriticalBlock block(sect);
+        // to handle add(A), add(B), remove(A), remove(B), add(A), add(B)
+        // and try to preserve add/remove order we use two loops
+        bool has_del = false;
         unsigned n = items.ordinality();
-        for (unsigned i=0;i<n;)
+        for (unsigned i=0;i<n;i++)
         {
             SelectItem *si = items.element(i);
             if (si->del)
             {
+                has_del = true;
                 epoll_op(epfd, EPOLL_CTL_DEL, si, 0);
-                // Release/dtors should not throw but leaving try/catch here until all paths checked
+            }
+            else if (si->add_epoll)
+            {
+                si->add_epoll = false;
+                if (si->mode != 0)
+                {
+                    unsigned int ep_mode = 0;
+                    if (si->mode & SELECTMODE_READ)
+                        ep_mode |= EPOLLIN;
+                    if (si->mode & SELECTMODE_WRITE)
+                        ep_mode |= EPOLLOUT;
+                    if (si->mode & SELECTMODE_EXCEPT)
+                        ep_mode |= EPOLLPRI;
+                    if (ep_mode != 0)
+                        epoll_op(epfd, EPOLL_CTL_ADD, si, ep_mode);
+                }
+            }
+        }
+
+        if (!has_del)
```

Comment:
If instead of boolean, you recorded highest 'i' delete in 1st pass, then 2nd pass could stop at that point rather than process whole of 'n'.

Reply:
Alternatively 1st pass could remove and swap items as before, but proceed to next item and record position if swapped in item is new.
Then on 2nd pass, if there's a recorded position, it is the high point it needs to look at (for add_epoll==true items).

It would mean 1st pass reduced items array size as well as reducing # of items to iterate over in 2nd pass.